### PR TITLE
Wrap timestamps in ROWTIME expressions with STRINGTOTIMESTAMP

### PR DESCRIPTION
**Pull request by jzaralim**
### Description 
Because the `ROWTIME` is stored as a Long, the user either has to pass in Long values or use  `STRINGTOTIMESTAMP` function to make queries using `ROWTIME`, for example:

`ROWTIME > STRINGTOTIMESTAMP('1970-01-01', 'yyyy-mm-dd')`
`ROWTIME BETWEEN 156469222000 AND 1564692253275`

This change wraps datestrings with `STRINGTOTIMESTAMP` so the user can directly pass datestrings for querying:

`ROWTIME > '1970-01-01'`
`ROWTIME BETWEEN '1970-01-01 08:00:00' AND ''1970-01-01 10:00:00`

### Testing done 
Wrote tests for several types of expressions.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/syntax-reference.rst [hunk 1]
```diff
@@ -1153,6 +1153,17 @@ Example:
       WHERE ROWTIME >= 1510923225000
         AND ROWTIME <= 1510923228000;
 
+When writing logical expressions using `ROWTIME`, ISO-8061 formatted datestrings can also be used to represent dates.
```
Comment (JimGalasyn) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
When writing logical expressions using ``ROWTIME``, ISO-8061 formatted datestrings can also be used to represent dates.
```

File: docs/developer-guide/syntax-reference.rst [hunk 2]
```diff
@@ -1153,6 +1153,17 @@ Example:
...
+For example, the above query is equivalent to the following:
+
+.. code:: sql
+    SELECT * FROM pageviews
```
Comment (JimGalasyn) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
    
    SELECT * FROM pageviews
```

File: docs/developer-guide/syntax-reference.rst [hunk 3]
```diff
@@ -1153,6 +1153,17 @@ Example:
...
+          WHERE ROWTIME >= '2017-11-17 04:53:45'
+            AND ROWTIME <= '2017-11-17 04:53:48';
+
+If the datestring is inexact, the rest of the timestamp is assumed to be 0.
+For example, `ROWTIME = `2019-07-30 11:00` is equivalent to `ROWTIME = `2019-07-30 11:00:00`.
```
Comment (JimGalasyn) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
For example, ``ROWTIME = 2019-07-30 11:00`` is equivalent to ``ROWTIME = 2019-07-30 11:00:00``.
```

File: docs/developer-guide/syntax-reference.rst (hunk 1 above)
Comment (jzaralim) [inline docs/developer-guide/syntax-reference.rst]: 8601

File: docs/developer-guide/syntax-reference.rst (hunk 3 above)
Comment (blueedgenick) [reply docs/developer-guide/syntax-reference.rst]: i might explicitly call out here that all unspecified elements of the full pattern are right-padded with zeros

Comment (blueedgenick) [reply docs/developer-guide/syntax-reference.rst]: somewhere in here i would expect to see a mention of timezone handling too, as this always trips people up.... i.e. in what TZ are your iso-8601 strings interpreted ? that of teh running jvm or utc or ....?

**Comment by blueedgenick** [comment]
can we feature-flag this somewhere, so that we can easily turn it off once we have proper datetime data-type support ? i assume that we will eventually change the datatype of the ROWTIME pseudo-column to `dateitme`, once such a thing exists, and this kind of "auto-cast" can then get terribly confusing

File: docs/developer-guide/syntax-reference.rst [hunk 4]
```diff
@@ -1153,6 +1153,20 @@ Example:
...
+.. code:: sql
+
+    SELECT * FROM pageviews
+          WHERE ROWTIME >= '2017-11-17 04:53:45'
+            AND ROWTIME <= '2017-11-17 04:53:48';
+
+If the datestring is inexact, the rest of the timestamp is assumed to be padded with 0's.
+For example, ``ROWTIME = '2019-07-30 11:00'`` is equivalent to ``ROWTIME = '2019-07-30 11:00:00.0000'``.
+
+Note that the timestamps are interperted in the UTC timezone.
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]: UTC isn't going to work for everyone. We should support timezone information in the datetime string, but default to UTC is none is provided.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    final Query statement = (Query) KsqlParserTestUtil.buildSingleAst(query, metaStore).getStatement();
+    final Expression predicate = statement.getWhere().get();
+    final Expression rewritten = new StatementRewriteForRowtime(predicate).rewriteForRowtime();
+
+    assertThat(rewritten.toString(), equalTo("((ORDERS.ROWTIME > STRINGTOTIMESTAMP('2017-01-01 00:00:00.000', 'yyyy-MM-dd HH:mm:ss.SSS')) AND (ORDERS.ROWKEY = '2017-01-01'))"));
+  }
+
+  @Test
+  public void shouldHandleNestedExpressions() {
+    final String simpleQuery = "SELECT * FROM orders where FOO(ROWTIME + 6000) > '2017-01-01 00:00:00.000' + 35;";
```
Comment (big-andy-coates) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java]: Isn't this a bit dangerous? We're assuming FOO is returning a `long` and that represents a datetime, but this might not be the case, e.g. `SELECT * FROM orders where IS_LEAP_YEAR(ROWTIME + 6000) > '2017-01-01 00:00:00.000' + 35;` wouldn't make sense!

I don't think we can make any such assumption on the return value from a function.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    final Expression rewritten = new StatementRewriteForRowtime(predicate).rewriteForRowtime();
+
+    assertThat(rewritten.toString(), equalTo("(ORDERS.ROWTIME BETWEEN"
+        + " STRINGTOTIMESTAMP('2017-01-01 00:00:00.000', 'yyyy-MM-dd HH:mm:ss.SSS') AND"
+        + " STRINGTOTIMESTAMP('2017-02-01 00:00:00.000', 'yyyy-MM-dd HH:mm:ss.SSS'))"));
+  }
+
+  @Test
+  public void shouldNotProcessStringsInFunctions() {
+    final String query = "SELECT * FROM orders where ROWTIME = foo('bar');";
```
Comment (big-andy-coates) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java]: would a better test not be `SELECT * FROM orders where ROWTIME = foo('2017-01-01');`

**Comment by big-andy-coates** [comment]
@blueedgenick

> can we feature-flag this somewhere, so that we can easily turn it off once we have proper datetime data-type support

This is all hidden from the user anyway. So once we have a datetime type if/when we are able to flip `ROWTIME` to be of type `datetime` we'll need to make sure all the same comparisons work. So I don't think we need a feature flag, just to remember to tidy up old code once `ROWTIME` is a `datetime`.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtime.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+public class StatementRewriteForRowtime {
+  private final Expression expression;
+
+  public StatementRewriteForRowtime(final Expression expression) {
+    this.expression = Objects.requireNonNull(expression, "expression");
+  }
+
+  public static boolean requiresRewrite(final Expression expression) {
+    return expression.toString().contains("ROWTIME");
```
Comment (hjafarpour) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtime.java]: This will match any expression with `ROWTIME` in it. For instance, it will match the following expression where you should not indeed rewrite the expression:
`TIMESTAMPTOSTRING(ROWTIME, 'yyyy-MM-dd HH:mm:ss') = '2019-08-08 12:12:12'`.
You need to change this such that it only rewrites the cases you have mentioned in the docs.

Comment (jzaralim) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtime.java]: This particular statement remained the same, but the rewriter itself has been updated to only rewrite timestrings when one side of the comparison is exactly `ROWTIME`.

File: ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java (hunk 5 above)
Comment (jzaralim) [reply ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java]: My thoughts were that this is purely a query rewriter, so those kinds of cases would get caught later on. So for example, `IS_LEAP_YEAR(ROWTIME + 6000) > 35` passes this step, but would still eventually get caught during evaluation. I did change the overall logic of the rewriter as per @hjafarpour's comment though.

**Comment by blueedgenick** [comment]
@big-andy-coates 
>This is all hidden from the user anyway. So once we have a datetime type if/when we are able to flip ROWTIME to be of type datetime we'll need to make sure all the same comparisons work. So I don't think we need a feature flag, just to remember to tidy up old code once ROWTIME is a datetime.

this is kind of my point - what's really happening here is that we are introducing some limited form of "auto-casting" from one datatype (long-of-a-timestamp) to another (string). 
This kind of automated attempt to be helpful with datatypes rarely works out well in practice as it only obfuscates to the user what the true types of things are, is incomplete (not covering all possible conversions), and can have unpredictable or surprising consequences. I'd recommend against pursuing any further form of this pattern. You'll notice that very few, if any, rdbms products do this without forcing you to do a proper `cast()` or `convert()` in your sql.

**Comment by jzaralim** [comment]
Hey all, I realized that I've hit "merge" prematurely. If anyone has any more feedback, I can submit another PR. Sorry about this!

**Comment by big-andy-coates** [comment]
@blueedgenick ah, that's what you were getting at... then I whole heartily agree!

File: ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java (hunk 5 above)
Comment (big-andy-coates) [reply ksql-parser/src/test/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtimeTest.java]: They may get caught later on, but the UX for the user is not great.  There's going to get some message that `Foo can not be compared to Long` or something, which is very confusing for the user.

File: ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtime.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  }
+
+  private static class TimestampRewriter extends StatementRewriter {
+    private static final String DATE_PATTERN = "yyyy-MM-dd";
+    private static final String TIME_PATTERN = "HH:mm:ss.SSS";
+    private static final String PATTERN = DATE_PATTERN + "'T'" + TIME_PATTERN;
+
+    @Override
+    public Expression visitFunctionCall(final FunctionCall node, final Object context) {
+      return (Expression) new StatementRewriter().process(node, context);
```
Comment (big-andy-coates) [inline ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtime.java]: Does this line do anything other than return a duplicate of the function call node? I think `StatementRewriter` should be `abstract` as its meant as a base class for rewrites. In itself it doesn't actually change the query, right?

Comment (jzaralim) [reply ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/StatementRewriteForRowtime.java]: It stops the rewriter from further processing anything within a function. So, if the expression looks like `ROWTIME > FOO('2017-01-01 ' + 100)  + '2019-01-01'`, then ` '2019-01-01'` will get wrapped but `'2017-01-01'` doesn't.

**Comment by jzaralim** [comment]
@big-andy-coates The decision flow for a rewrite are as follows:

1. If a filter contains the string `ROWTIME` anywhere in it, pass it to the rewriter.
2. If the rewriter ever reaches a Comparision expression, check if one of the sides is exactly `ROWTIME`. If one of them is, pass its subnodes to another rewriter (`TimestampRewriter`).
3. In `TimestampRewriter`, if a string is reached, wrap it in `STRINGTOTIMESTAMP`. However, if a function is reached, don't process the arguments of it.

So, the following statements do not get rewritten:
`ROWTIME > FOO(5 + '2017-01-01')`
`FOO(ROWTIME) > '2017-01-01'`
`5 + ROWTIME>'2017-01-01'`

As for error handling, ideally `STRINGTOTIMESTAMP` would catch them. However, it seems that `STRINGTOTIMESTAMP` not catching parsing errors is not just a simple bug, so the next PR will have the following changes:

1. Instead of wrapping the datestrings, rewrite them by directly converting them to `Long`. That way, the errors caught by the datestring parser could be passed on to the user.
2. Add more tests to go over more edge cases.
